handler/graphql.go

```diff
@@ -390,187 +388,199 @@ func computeQueryHash(query string) string {
+			if ok {
+				doc = val.(*ast.QueryDocument)
+				cacheHit = true
+			}
+		}
+
+		ctx, doc, gqlErr := gh.parseOperation(ctx, &parseOperationArgs{
+			Query:     reqParams.Query,
+			CachedDoc: doc,
+		})
+		if gqlErr != nil {
+			sendError(w, http.StatusUnprocessableEntity, gqlErr)
 			return
 		}
-		if reqParams.Extensions.PersistedQuery.Version != 1 {
-			sendErrorf(w, http.StatusOK, "Unsupported persisted query version")
+
+		ctx, op, vars, listErr := gh.validateOperation(ctx, &validateOperationArgs{
+			Doc:           doc,
+			OperationName: reqParams.OperationName,
+			CacheHit:      cacheHit,
+			R:             r,
+			Variables:     reqParams.Variables,
+		})
+		if len(listErr) != 0 {
+			sendError(w, http.StatusUnprocessableEntity, listErr...)
 			return
 		}
-		if reqParams.Query == "" {
-			// client sent optimistic query hash without query string
-			query, ok := gh.cfg.apqCache.Get(ctx, queryHash)
-			if !ok {
-				sendErrorf(w, http.StatusOK, errPersistedQueryNotFound)
-				return
-			}
-			reqParams.Query = query
-		} else {
-			if computeQueryHash(reqParams.Query) != queryHash {
-				sendErrorf(w, http.StatusOK, "provided sha does not match query")
-				return
+
+		if gh.cache != nil && !cacheHit {
+			gh.cache.Add(reqParams.Query, doc)
+		}
+
+		reqCtx := gh.cfg.newRequestContext(gh.exec, doc, op, reqParams.Query, vars)
+		ctx = graphql.WithRequestContext(ctx, reqCtx)
+
+		defer func() {
+			if err := recover(); err != nil {
+				userErr := reqCtx.Recover(ctx, err)
+				sendErrorf(w, http.StatusUnprocessableEntity, userErr.Error())
 			}
-			apqRegister = true
+		}()
+
+		if gh.cfg.complexityLimitFunc != nil {
+			reqCtx.ComplexityLimit = gh.cfg.complexityLimitFunc(ctx)
 		}
-	} else if reqParams.Query == "" {
-		sendErrorf(w, http.StatusUnprocessableEntity, "Must provide query string")
-		return
-	}
-
-	var doc *ast.QueryDocument
-	var cacheHit bool
-	if gh.cache != nil {
-		val, ok := gh.cache.Get(reqParams.Query)
-		if ok {
-			doc = val.(*ast.QueryDocument)
-			cacheHit = true
+
+		if reqCtx.ComplexityLimit > 0 && reqCtx.OperationComplexity > reqCtx.ComplexityLimit {
+			sendErrorf(w, http.StatusUnprocessableEntity, "operation has complexity %d, which exceeds the limit of %d", reqCtx.OperationComplexity, reqCtx.ComplexityLimit)
+			return
 		}
-	}
 
-	ctx, doc, gqlErr := gh.parseOperation(ctx, &parseOperationArgs{
-		Query:     reqParams.Query,
-		CachedDoc: doc,
-	})
-	if gqlErr != nil {
-		sendError(w, http.StatusUnprocessableEntity, gqlErr)
-		return
-	}
+		if apqRegister && gh.cfg.apqCache != nil {
+			// Add to persisted query cache
+			gh.cfg.apqCache.Add(ctx, queryHash, reqParams.Query)
+		}
 
-	ctx, op, vars, listErr := gh.validateOperation(ctx, &validateOperationArgs{
-		Doc:           doc,
-		OperationName: reqParams.OperationName,
-		CacheHit:      cacheHit,
-		R:             r,
-		Variables:     reqParams.Variables,
+		switch op.Operation {
+		case ast.Query:
+			b, err := json.Marshal(gh.exec.Query(ctx, op))
```

Comment:
this json.Marshal might need to be pluggable too, maybe part of the same transport hook?

What would this look like if we wanted to use GRPC as a transport instead of json+POST?